src/bin/purge_ttl.rs

```diff
@@ -69,32 +76,192 @@ pub fn statsd_from_env() -> Result<StatsdClient, Box<dyn Error>> {
     session: &Session,
-) -> Result<ExecuteSqlRequest, Box<dyn Error>> {
+    request_type: RequestType,
+) -> Result<(ExecuteSqlRequest, Vec<u8>), Box<dyn Error>> {
     // Create a transaction
     let mut opt = TransactionOptions::new();
-    opt.set_partitioned_dml(TransactionOptions_PartitionedDml::new());
+    match request_type {
+        RequestType::ReadWrite => {
+            opt.set_read_write(TransactionOptions_ReadWrite::new());
+        }
+        RequestType::ReadOnly => {
+            opt.set_read_only(TransactionOptions_ReadOnly::new());
+        }
+        RequestType::PartitionedDml => {
+            opt.set_partitioned_dml(TransactionOptions_PartitionedDml::new());
+        }
+    }
+
     let mut req = BeginTransactionRequest::new();
     req.set_session(session.get_name().to_owned());
     req.set_options(opt);
     let mut txn = client.begin_transaction(&req)?;
 
+    let id = txn.take_id();
     let mut ts = TransactionSelector::new();
-    ts.set_id(txn.take_id());
+    ts.set_id(id.clone());
 
-    // Create an SQL request
     let mut req = ExecuteSqlRequest::new();
     req.set_session(session.get_name().to_string());
     req.set_transaction(ts);
 
+    Ok((req, id))
+}
+
+fn continue_transaction(
+    session: &Session,
+    transaction_id: Vec<u8>,
+) -> Result<ExecuteSqlRequest, Box<dyn Error>> {
+    let mut ts = TransactionSelector::new();
+    ts.set_id(transaction_id);
+    let mut req = ExecuteSqlRequest::new();
+    req.set_session(session.get_name().to_string());
+    req.set_transaction(ts);
     Ok(req)
 }
 
+fn commit_transaction(
+    client: &SpannerClient,
+    session: &Session,
+    txn: Vec<u8>,
+) -> Result<(), Box<dyn Error>> {
+    let mut req = CommitRequest::new();
+    req.set_session(session.get_name().to_owned());
+    req.set_transaction_id(txn);
+    client.commit(&req)?;
+    Ok(())
+}
+
+pub struct SyncResultSet {
+    result: googleapis_raw::spanner::v1::result_set::ResultSet,
+}
+
+impl Iterator for SyncResultSet {
+    type Item = Vec<Value>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let rows = &mut self.result.rows;
+        if rows.is_empty() {
+            None
+        } else {
+            let row = rows.remove(0);
+            Some(row.get_values().to_vec())
+        }
+    }
+}
+
+fn delete_incremental(
+    client: &SpannerClient,
+    session: &Session,
+    table: String,
+    column: String,
+    chunk_size: u64,
+) -> Result<(), Box<dyn Error>> {
+    let (mut req, mut txn) = begin_transaction(&client, &session, RequestType::ReadWrite)?;
+    let select_sql = format!("SELECT fxa_uid, fxa_kid, collection_id, {} FROM {} WHERE expiry < CURRENT_TIMESTAMP() LIMIT {}", column, table, chunk_size);
+    trace!("Selecting rows to delete: {}", select_sql);
+    req.set_sql(select_sql.clone());
+    let mut result = SyncResultSet {
+        result: client.execute_sql(&req)?,
+    };
+
+    let mut total: usize = 1;
+    while total < 1_000_000 {
```

Comment:
This select could also probably live inside of a `loop {}` (with an extra break for total >= 1_000_000) to avoid duplicating it. Basically emulating a `do while` style loop

Reply:
Great suggestion.